c++/src/capnp/serialize-async.h

```diff
@@ -139,6 +141,166 @@ class AsyncCapabilityMessageStream final: public MessageStream {
+      kj::ArrayPtr<const int> fds,
+      kj::ArrayPtr<const kj::ArrayPtr<const word>> segments) override;
+  kj::Promise<void> writeMessages(
+      kj::ArrayPtr<kj::ArrayPtr<const kj::ArrayPtr<const word>>> messages) override;
+  kj::Maybe<int> getSendBufferSize() override;
+  kj::Promise<void> end() override;
+
+private:
+  kj::AsyncIoStream& stream;
+  BufferedStreamOptions options;
+
+  struct IncompleteRead {
+    kj::Array<kj::byte> bytes;
+    size_t expectedSize;
+  };
+
+  std::deque<kj::Own<MessageReader>> bufferedReaders;
+  kj::Maybe<IncompleteRead> incompleteRead;
+
+  // Implementation details used by tryReadMessageInternal
+  kj::Maybe<MessageReaderAndFds> tryGetBufferedMessage(kj::ArrayPtr<kj::AutoCloseFd> fdSpace);
+  kj::Maybe<IncompleteRead> tryGetIncompleteRead();
+
+  struct MessageReaderAndOwnedFds {
+    kj::Own<MessageReader> reader;
+    kj::Array<kj::AutoCloseFd> fds;
+  };
+
+  void bufferCompleteMessage(MessageReaderAndOwnedFds message);
+  void bufferIncompleteMessage(kj::ArrayPtr<kj::byte> leftover, kj::Array<kj::AutoCloseFd> fds, size_t size);
+
+  struct ReadResult {
+    kj::Array<kj::byte> buffer;
+    size_t byteCount;
+
+    kj::Array<kj::AutoCloseFd> getFds(bool isLast) { return nullptr; }
+  };
+  kj::Promise<ReadResult> tryRead(kj::Array<byte> buffer, void* start, size_t minBytes,
+      size_t maxBytes);
+
+  friend kj::Promise<kj::Maybe<MessageReaderAndFds>>
+      _::tryReadMessageInternal<BufferedAsyncIoMessageStream>(
+      BufferedAsyncIoMessageStream& stream,
+      uint64_t defaultReadSize, kj::ArrayPtr<kj::AutoCloseFd> passedInFdSpace,
+      ReaderOptions readerOptions, kj::ArrayPtr<word> scratchSpace);
+};
+
+class BufferedAsyncCapabilityMessageStream final : public MessageStream {
+public:
+  explicit BufferedAsyncCapabilityMessageStream(kj::AsyncCapabilityStream& stream,
+      BufferedCapabilityStreamOptions options = BufferedCapabilityStreamOptions{})
+      : stream(stream), options(options) {}
+
+  // Implements MessageStream
+  kj::Promise<kj::Maybe<MessageReaderAndFds>> tryReadMessage(
+      kj::ArrayPtr<kj::AutoCloseFd> fdSpace,
+      ReaderOptions options = ReaderOptions(), kj::ArrayPtr<word> scratchSpace = nullptr) override;
+  kj::Promise<void> writeMessage(
+      kj::ArrayPtr<const int> fds,
+      kj::ArrayPtr<const kj::ArrayPtr<const word>> segments) override;
+  kj::Promise<void> writeMessages(
+      kj::ArrayPtr<kj::ArrayPtr<const kj::ArrayPtr<const word>>> messages) override;
+  kj::Maybe<int> getSendBufferSize() override;
+  kj::Promise<void> end() override;
+
+private:
+  kj::AsyncCapabilityStream& stream;
+  BufferedCapabilityStreamOptions options;
+
+  struct MessageReaderAndOwnedFds {
+    kj::Own<MessageReader> reader;
+    kj::Array<kj::AutoCloseFd> fds;
+  };
+
+  std::deque<MessageReaderAndOwnedFds> bufferedReaders;
+
+  struct IncompleteRead {
+    kj::Array<kj::byte> bytes;
+    size_t expectedSize;
+  };
+
+  kj::Maybe<IncompleteRead> incompleteRead;
```

Comment:
Can you add some comments here explaining what all these pieces of state represent?

It's hard to follow the body of `tryReadMessageInternal` without first understanding the starting state.